Add contributing guidelines (#35)
GRAAL-Research/poutyne: CONTRIBUTING.md

# Contributing to Poutyne
We love your input! We want to make contributing to this project as easy and transparent as possible, whether it's:

- Reporting a bug
- Discussing the current state of the code
- Submitting a fix
- Proposing new features
- Becoming a maintainer

## We Develop with Github
We use github to host code, to track issues and feature requests, as well as accept pull requests.

## We Use [Github Flow](https://guides.github.com/introduction/flow/index.html), So All Code Changes Happen Through Pull Requests
Pull requests are the best way to propose changes to the codebase. We actively welcome your pull requests:

> 1. Fork the repo and create your branch from `master`.

Reviewer: Should we propose a branch name pattern ?
Something like: `fixing_bug-... `

Author: Since we suggest to fork the repository, the contributor never creates a branch on the Poutyne git, only on his personal workspace. When submitting the PR, he then proposes changes coming from his own fork and therefore branch naming is not really an issue here.

Correct me if i'm wrong @freud14 .

> 2. If you've added code that should be tested, add tests.

Reviewer: If you've added new code, then it **must** be tested. 
­

Author: ```suggestion
2. If you've added code that should be tested, you **must** ensure it is properly tested.
```

3. If you've changed APIs, update the documentation.

> 4. Ensure the test suite passes.

Reviewer: Should not we have a travis test suite build up for that ? (removing the concept to our automation system)

Author: We already have a Travis set up. It's more of a "please ensure the Travis tests did not fail".

Author: ```suggestion
4. Ensure the Travis test suite passes.
```

> 5. Make sure your code lints.

Reviewer: Should not we use a linter in the .yml of travis ? (easier and more consistent with our style)

Author: I'm no expert with Travis. Can we automatically run the linting and/or the formatting with Travis. 

Let's say we add the linting, how does the contributor see the results ?

Reviewer 2: We probably can run the linting in Travis but I don't know if there is a simple way to make it fail the tests if linting does not pass. Also, if we do that, the pylint version should be fixed to some version because each version tends to add warnings that were not there previously.

With yapf, is it possible to just test the formatting and not modify the files? If so, it can be a good idea to add it to the Travis script.

Anyway, I don't think that these modifications fall in the scope of this pull request.

Author: Agreed, the Travis integration is interesting but out of scope here.

Reviewer: I'll take a look for the linter integration also, just to be sure we don't miss any bad syntax.

> 6. Issue that pull request!

Reviewer: Is it necessary ?

Author: Issue that PR => submit it. I'll change the name.

Author: ```suggestion
6. Submit that pull request!
```

Reviewer: Oh I thought it was the concept of issue.

## Any contributions you make will be under the MIT Software License
In short, when you submit code changes, your submissions are understood to be under the same [MIT License](http://choosealicense.com/licenses/mit/) that covers the project. Feel free to contact the maintainers if that's a concern.

## Report bugs using Github's [issues](https://github.com/GRAAL-Research/poutyne/issues)
We use GitHub issues to track public bugs. Report a bug by [opening a new issue](); it's that easy!

> ## Write bug reports with detail, background, and sample code

Reviewer: *###

Author: Why the third '#' ? I feel like it is okay for this title to be at same emphasis level as 'Use a Consistent Coding Style' and family.

Reviewer: The third will put the concept into the title report bugs...

Author: Have a look at 649a7bf483f1742f57c0ebf446ce449b4c88d159, I think it clears the confusion there.

Reviewer: Better

**Great Bug Reports** tend to have:

> - A quick summary and/or background

Reviewer: and clear

Author: ```suggestion
- A quick and clear summary and/or background
```

- Steps to reproduce

>   - Be specific!

Reviewer: and clear

Author: ```suggestion
  - Be specific and clear!
```

  - Give sample code if you can. Try to reduce the bug to the minimum amount of code needed to reproduce: it will help in our troubleshooting procedure.
- What you expected would happen
- What actually happens
- Notes (possibly including why you think this might be happening, or stuff you tried that didn't work)

> 

Reviewer: Feel free ton include print screen also for more clarity.

Author: ```suggestion
Feel free to include any print screen or other file you feel may further clarify your point.
```

> ## Use a Consistent Coding Style

Reviewer: ```suggestion
## Do you have a suggestion for an enhancement?

We use GitHub issues to track enhancement requests.  Before you create an enhancement request:

* Make sure you have a clear idea of the enhancement you would like.  If you have a vague idea, consider discussing
it first on the users list.

* Check the documentation to make sure your feature does not already exist.

* Do [a quick search](https://github.com/allenai/allennlp/issues) to see whether your enhancement has already been suggested.

When creating your enhancement request, please:

* Provide a clear title and description.

* Explain why the enhancement would be useful.  It may be helpful to highlight the feature in other libraries.

* Include code examples to demonstrate how the enhancement would be used.

## Use a Consistent Coding Style
```

Reviewer: From [AllenNLP](https://github.com/allenai/allennlp/blob/master/CONTRIBUTING.md)

All of the code is formatted using [yapf](https://github.com/google/yapf) with the associated [config file](https://github.com/GRAAL-Research/poutyne/blob/master/.style.yapf). In order to format the code of your submission, simply run

```
yapf poutyne --recursive --in-place
```

We also have our own `pylint` [config file](https://github.com/GRAAL-Research/poutyne/blob/master/.pylintrc). Try not to introduce code incoherences detected by the linting. You can run the linting procedure with

```
pylint poutyne
```

> 

Reviewer: As stated before, should not this step been include into our build phase ?

## Tests

If your pull request introduces a new feature, please deliver it with tests that ensure correct behavior. All of the current tests are located under the `tests` folder, if you want to see some examples. 

For any pull request submitted, **ALL** of the tests must succeed. You can run the tests with

```
pytest tests
```

> 

Reviewer: Suggestion from [AllenNLP How To](https://github.com/allenai/allennlp/blob/master/CONTRIBUTING.md)
Do you have a suggestion for an enhancement?

We use GitHub issues to track enhancement requests. Before you create an enhancement request:

    Make sure you have a clear idea of the enhancement you would like. If you have a vague idea, consider discussing it first on the users list.

    Check the documentation to make sure your feature does not already exist.

    Do a quick search to see whether your enhancement has already been suggested.

When creating your enhancement request, please:

    Provide a clear title and description.

    Explain why the enhancement would be useful. It may be helpful to highlight the feature in other libraries.

    Include code examples to demonstrate how the enhancement would be used.
